DataDog/dd-trace-rb
# [PROF-8917] Add support for the libdatadog crash tracker

**What does this PR do?**

This PR adds support for the libdatadog crash tracker feature (off by default).

The crash tracker works by detecting when the Ruby VM reaches a segmentation fault, reporting the crash information as a last profile before the VM dies.

All of the interesting work is in <https://github.com/DataDog/libdatadog/pull/282>, this PR basically just wires things up.

**Motivation:**

This will be a useful tool when debugging VM crashes.

**Additional Notes:**

I'm opening this PR as a draft as the libdatadog support has not yet landed/been released. Also, there's a few open questions on:

* fork handling
* when to shut down

**How to test the change?**

(You'll need to build <<https://github.com/DataDog/libdatadog/pull/282> until the crash tracker gets included in a libdatadog release)

To test the crash tracker with an actual crash, try running the following on Ruby 2.6:

```bash
$ DD_PROFILING_ENABLED=true DD_PROFILING_EXPERIMENTAL_CRASH_TRACKING_ENABLED=true DD_TRACE_DEBUG=true bundle exec ddtracerb exec ruby -e 'Process.detach(fork { exit! }).instance_variable_get(:@foo)'
```

This should also work in the future but right now it doesn't work correctly; still looking into why:

```bash
$ DD_PROFILING_ENABLED=true DD_PROFILING_EXPERIMENTAL_CRASH_TRACKING_ENABLED=true DD_TRACE_DEBUG=true bundle exec ddtracerb exec ruby -e 'Process.kill("SEGV", Process.pid)'
```

**For Datadog employees:**
credentials of any kind, I've requested a review from `@DataDog/security-design-and-guidance`.
